### PR TITLE
feat: support render svg in <img/>

### PR DESCRIPTION
**Pull request by XGHeaven**
The core changes are:

1. Refactor the `` render process. (very important changes)
    1. Now support using `Image` or `RenderObject` as `RenderReplaced`'s child. You can use `_updateRenderObject` function to change it.
    1. When `` disconnected from dom tree, no more to dropChild or dispose child. Instead of, dispose child when `` disposed.
    1. Improve the performance for rendering image. (I think. No perf test)
    1. And others
1. Expose the Gumbo parser function from C side to dart side for parsing the SVG source.
1. Move the code that is setting the value to `renderStyle` from `element.dart` to `render_style.dart`.

Preview: create-react-app demo

image

### CODE REVIEW

File: webf/lib/src/bridge/native_gumbo.dart [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  print('$p node type: $type');
+  switch(type) {
+    case 1: {
+      // element
+      final element = nodePtr.ref.v.element;
+      final children = element.children;
+      final tag = element.tag;
+      final tagName = element.original_tag.length != 0 ? element.original_tag.data.toDartString(length: element.original_tag.length) : '(none)';
+
+      print('$p tag: $tag($tagName) in namespace ${element.tag_namespace} with ${children.length} child');
```
Comment (andycall) [inline webf/lib/src/bridge/native_gumbo.dart:135]: remove print

File: webf/lib/src/svg/svg_render_object_builder.dart [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    });
+
+    if (root == nullptr) {
+      // IMPROVE: throw a more specific error
+      throw Error();
+    }
+
+    final rootRenderObject = visitSVGTree(root, (node, parent) {
+      final type = node.ref.type;
+      if (type == 1) {
```
Comment (andycall) [inline webf/lib/src/svg/svg_render_object_builder.dart:74]: enum values

File: webf/lib/src/bridge/native_gumbo.dart [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  external NativeGumboSourcePosition value_end;
+}
+
+// DEBUG_FUNCTION
+void printGumboNodeTree(Pointer<NativeGumboNode> nodePtr, [int indent = 0]) {
+  final p = ' ' * (indent * 2);
+  final type = nodePtr.ref.type;
+  print('$p node type: $type');
+  switch(type) {
+    case 1: {
```
Comment (andycall) [inline webf/lib/src/bridge/native_gumbo.dart:128]: what's meaning of `1`?

File: webf/lib/src/painting/box_fit_image.dart [hunk 4]
```diff
@@ -116,15 +123,15 @@ class BoxFitImage extends ImageProvider<BoxFitImageKey> {
...
+            () => stream.completer!,
         onError: handleError,
       );
       assert(identical(completer, stream.completer));
       return;
     }
     final ImageStreamCompleter? completer = PaintingBinding.instance.imageCache.putIfAbsent(
       key,
-      () => loadImage(key, PaintingBinding.instance.instantiateImageCodecWithSize),
+          () => loadImage(key, PaintingBinding.instance.instantiateImageCodecWithSize),
```
Comment (andycall) [inline webf/lib/src/painting/box_fit_image.dart:134]: code format

File: webf/lib/src/bridge/native_gumbo.dart (hunk 1 above)
Comment (XGHeaven) [reply webf/lib/src/bridge/native_gumbo.dart:135]: This is a debug function which is not called in production. So `print` is necessary.

File: webf/lib/src/html/img.dart [hunk 5]
```diff
@@ -16,6 +16,8 @@ import 'package:webf/foundation.dart';
 import 'package:webf/painting.dart';
 import 'package:webf/rendering.dart';
 
+import '../svg/svg_render_object_builder.dart';
```
Comment (andycall) [inline webf/lib/src/html/img.dart:19]: import 'package:webf/svg.dart';

File: webf/lib/src/html/img.dart [hunk 6]
```diff
@@ -374,10 +450,6 @@ class ImageElement extends Element {
   void _stopListeningStream({bool keepStreamAlive = false}) {
     if (!_isListeningStream) return;
 
-    if (keepStreamAlive && _completerHandle == null && _cachedImageStream?.completer != null) {
```
Comment (andycall) [inline webf/lib/src/html/img.dart:377]: Why this line are removed

File: webf/lib/src/bridge/native_gumbo.dart [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+
+  external NativeGumboSourcePosition name_start;
+  external NativeGumboSourcePosition name_end;
+
+  external NativeGumboSourcePosition value_start;
+  external NativeGumboSourcePosition value_end;
+}
+
+// DEBUG_FUNCTION
+void printGumboNodeTree(Pointer<NativeGumboNode> nodePtr, [int indent = 0]) {
```
Comment (andycall) [inline webf/lib/src/bridge/native_gumbo.dart:294]: rename to debugPrintGumboNodeTree

File: webf/lib/src/rendering/replaced.dart [hunk 8]
```diff
@@ -33,6 +33,7 @@ class RenderReplaced extends RenderBoxModel with RenderObjectWithChildMixin<Rend
   bool _isInLazyRendering = false;
   bool get isInLazyRendering => _isInLazyRendering;
   set isInLazyRendering(bool value) {
+    print('set $this');
```
Comment (andycall) [inline webf/lib/src/rendering/replaced.dart:36]: remove print

File: webf/lib/src/svg/svg_render_object_builder.dart [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+import 'package:webf/css.dart';
+import 'package:webf/dom.dart';
+import 'package:webf/painting.dart';
+
+import 'package:webf/rendering.dart';
+import 'package:webf/src/svg/rendering/container.dart';
+import 'package:webf/src/svg/rendering/root.dart';
+import 'package:webf/svg.dart';
+
+import '../bridge/native_gumbo.dart';
```
Comment (andycall) [inline webf/lib/src/svg/svg_render_object_builder.dart:16]: import 'package:webf/bridge.dart';

File: webf/lib/src/html/img.dart (hunk 6 above)
Comment (XGHeaven) [reply webf/lib/src/html/img.dart:377]: I used another way to keep the image alive.

In the previous, the listener will remove when the render object detach. So must be to keep the image stream alive when it attached back.
I move those to dispose lifecycle, So don't keep the stream alive.

File: webf/lib/src/svg/svg_render_object_builder.dart [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    freeSVGResult(gumbo);
+
+    return rootRenderObject as RenderBoxModel;
+  }
+
+  RenderBoxModel getSVGRenderBox(String tagName) {
+    final Constructor = svgElementsRegistry[tagName];
+    if (Constructor != null) {
+      final element = Constructor(null);
+      // element.ownerDocument = target.ownerDocument;
```
Comment (andycall) [inline webf/lib/src/svg/svg_render_object_builder.dart:107]: ？

**Review by andycall — APPROVED** [review]

File: webf/lib/src/svg/svg_render_object_builder.dart (hunk 10 above)
Comment (XGHeaven) [reply webf/lib/src/svg/svg_render_object_builder.dart]: Removed

**Review by yifei8 — APPROVED** [review]

**Pull request merged**